docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+
+```
+// Get the value or a default.
+let x = opt_value.unwrap_or(other_value);
+```
+
+```
+// Get the value or return an Err
+// (this can only be called in a function that returns a Result)
+let y = opt_value.ok_or("Error message")?;
+```
+
+```
+// Transform with lazy default value (note how "into()" can elide the
+// type since it's already known to be OtherType).
+let z = opt_value.map_or_else(OtherType::get_default, |v| v.into());
+```
+
+## Iteration
+
+### Strongly prefer functional iteration :: suggested
+
+Prefer using iterators rather than directly indexing data to avoid the need for any explicit bounds checks. The Rust compiler will make this at least as performant as a manual loop.
+
+```
+let tags: Vec<Tag> = ["en", "fr", "de"].iter().map(Tag::from).collect();
+```
+
+```
+let mut tags: Vec<String> = Vec::new();
+let lang = ["en", "fr", "de"];
+for n in 0 .. lang.len() {
+    tags.push(Tag::from(lang[n]));
+}
+```
+
+## Enums
+
+### Strongly prefer enums to define states :: suggested
+
+Enums in Rust are cheap/free, and incredibly useful. They can be used (as in C++/Java) for providing named, bounded "choices", including avoiding bare boolean parameters, but they can also provide the basis for type-safe patterns such as [elegant finite state machines](https://bluejekyll.github.io/blog/fsm/rust/2015/08/13/rust-and-the-most-elegant-fsm.html) using generified enums with data.
+
+It's probably worth noting here that the [Result](https://doc.rust-lang.org/std/result/) type itself is just a normal enum in Rust with two values (`Ok` and `Err`).
+
+## Matching
+
+### Prefer match statements for exhaustive conditional code :: suggested
+
+Matching is Rust's idiomatic way to handle any non-trivial exhaustive conditional code, and works elegantly with enums, but handles any data for which a predicate can be formed (which is basically everything).
+
+Even a simple if/else block can be expressed more idiomatically using a match statement and will produce essentially the same compiled code after optimization. This is especially true if you want all cases to be covered, since this is enforced by a match statement.
+
+From [Rust By Example](https://doc.rust-lang.org/rust-by-example/flow_control/match.html):
+
+```
+match number {
+    // Match a single value
+    1 => println!("One!"),
+    // Match several values
+    2 | 3 | 5 | 7 | 11 => println!("This is a prime"),
+    // Match an inclusive range
+    13..=19 => println!("A teen"),
+    // Handle the rest of cases
+    _ => println!("Ain't special"),
+}
+```
+
+Another nice property of match is that it returns a value, which tied to the fact that it must cover all cases means you get a readable idiom for ad-hoc matching and mapping:
```

Comment:
Perhaps it may be interesting to note that:

match num % 2 {
  // ...
}

won't work.